### PR TITLE
Set security to agent tools

### PR DESCRIPTION
**Pull request by adrinr**
## Description
Set security to agent tools

## Screenshots
https://jam.dev/c/568c9232-446a-4db2-a5da-8abfe534301b

## Launchcontrol
Set security to agent tools

---
## Summary by cubic
Adds per‑tool execution security for agent tools with delegated “Run as” identities, strict call‑time authorization with auditing and per‑item result filtering, plus a role‑based chat preview for testing; gated by `AI_AGENT_TOOL_SECURITY` (default off).

- **New Features**
  - Execution policy: tools are “admin” or “configurable” (default principal is requester). The server resolves the effective principal per call, authorizes with the requester or preview role, and logs authorization decisions and execution outcomes.
  - Call‑time enforcement: tools stay available in prompts, but are authorized right before running. Collection results are filtered item‑by‑item by resource permissions. Denied calls return safe errors and the prompt tells the model not to substitute tools or claim success.
  - Builder: per‑tool “Run as” selector for configurable tools. Operation configs store tools as `[ { toolName, executionPrincipal } ]`. Tool renames preserve each tool’s principal.
  - Role‑based chat preview: pick any role (including Public) when testing. Roles refresh on open. Preview is allowed only for builders in development workspaces. The server rejects `previewRoleId` outside preview mode.
  - Permissions: requester authorization runs before execution. Query tools require WRITE; row tools require READ for get/search and WRITE for create/update; automation triggers check EXECUTE; escalation calls authorize with the full agent execution context. Admin execution is denied if the requester is now Public. Runs without a requester preserve prior behavior.

- **Migration**
  - No manual steps. Existing agents auto‑migrate `enabledTools` to `[ { toolName, executionPrincipal: "admin" } ]` to preserve prior behavior.
  - API: when creating/updating operations, send `enabledTools` as `[ { toolName, executionPrincipal } ]`.

<sup>Written for commit 9733f4e9d3f4f7c9cd2b742a57acc5440e9f0238. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Budibase/budibase/pull/19462?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/server/src/ai/tools/budibase/rows.ts [hunk 1]
```diff
@@ -399,6 +402,18 @@ export const createRowTools = ({
       sourceLabel: resolvedSourceLabel,
       sourceIconType,
       description,
+      authorization: {
+        supportedPrincipals: [
+          ToolExecutionPrincipal.REQUESTER,
+          ToolExecutionPrincipal.ADMIN,
+        ],
+        permissionType: PermissionType.TABLE,
+        permissionLevel:
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/server/src/ai/tools/budibase/rows.ts:443]: Solved

File: packages/server/src/ai/tools/budibase/rows.ts [hunk 2]
```diff
@@ -431,6 +434,20 @@ export const createRowTools = ({
...
+          action === "create_row" ||
+          action === "update_row" ||
+          action === "delete_row"
+            ? PermissionLevel.WRITE
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/server/src/ai/tools/budibase/rows.ts]: Better to keep it just in case it is added in the future

File: packages/server/src/ai/tools/authorization.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  if (!result || typeof result !== "object" || !(collectionKey in result)) {
+    return result
+  }
+  const collection = Reflect.get(result, collectionKey)
+  if (!Array.isArray(collection)) {
+    return result
+  }
+
+  const allowedItems = await Promise.all(
+    collection.map(async item => {
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/server/src/ai/tools/authorization.ts:37]: This issue does not seem longer valid

File: packages/server/src/sdk/workspace/ai/tests/run.ts [hunk 4]
```diff
@@ -289,7 +289,7 @@ async function runAgentForCase({
     selectedOperationName: run.selectedOperation?.name,
     promptInstructions: run.selectedOperation?.promptInstructions,
     enabledTools: run.selectedOperation?.enabledTools
-      ? [...run.selectedOperation.enabledTools]
+      ? run.selectedOperation.enabledTools.map(tool => tool.toolName)
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/server/src/sdk/workspace/ai/tests/run.ts:292]: Fixed

File: packages/builder/src/pages/builder/workspace/[workspaceId]/agent/[agentId]/config.svelte [hunk 5]
```diff
@@ -2,6 +2,7 @@
   import { Body, notifications, Select, Button } from "@budibase/bbui"
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/builder/src/pages/builder/workspace/[workspaceId]/agent/[agentId]/config.svelte:2]: Not introduced in this branch

[the remark this answers is not in the document]
Comment (adrinr) [reply packages/builder/src/pages/builder/workspace/[workspaceId]/agent/[agentId]/config.svelte:2]: Not introduced in this branch

File: packages/server/src/sdk/workspace/ai/agents/agentRuntime.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import { cache, features } from "@budibase/backend-core"
+import { cache, context, features } from "@budibase/backend-core"
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/server/src/sdk/workspace/ai/agents/agentRuntime.ts:1]: Not introduced in this branch

File: packages/server/src/sdk/workspace/ai/agents/utils.ts [hunk 7]
```diff
@@ -7,6 +7,8 @@ import {
   WebSearchProvider,
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/server/src/sdk/workspace/ai/agents/utils.ts:7]: This would add extra complexity for an edge case

File: packages/server/src/sdk/workspace/ai/tests/utils.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import { ToolExecutionPrincipal } from "@budibase/types"
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/server/src/sdk/workspace/ai/tests/utils.ts:1]: Done

File: packages/server/src/sdk/workspace/ai/agents/utils.ts [hunk 9]
```diff
@@ -197,14 +272,81 @@ export async function buildPromptAndTools(
...
+    const executionContext: AgentExecutionContext = {
+      tenantId: context.getTenantId(),
+      workspaceId,
+      agentId,
+      operationId: operation.id,
+      conversationId: options.execution.sessionId,
+      requestingUserId: options.execution.requestingUserId,
+      requestingUserIsPublic: options.execution.requestingUserIsPublic,
+    }
+    for (const tool of enabledTools) {
```
[the remark this answers is not in the document]
Comment (adrinr) [reply packages/server/src/sdk/workspace/ai/agents/utils.ts:213]: This would add extra complexity for an edge case

**Review by deanhannigan — APPROVED** [review]
LGTM! I gave this a good run through locally and it was really solid! The permission select on the chat preview is a fantastic touch 🚀
